File: scripts/check-rust-release.ts
#!/usr/bin/env bun
/** Bind release metadata to workspace sources, an exact tag, and archive bytes. */
import { createHash } from "node:crypto";
import { createReadStream, readFileSync } from "node:fs";
import { resolve } from "node:path";
import { parseArgs } from "node:util";

export const TARGET = "x86_64-unknown-linux-gnu";
export async function digest(path: string): Promise<string> {
  const hash = createHash("sha256");
  for await (const chunk of createReadStream(path)) hash.update(chunk);
  return hash.digest("hex");
}
function git(root: string, ref: string): string {
  const result = Bun.spawnSync(
    ["git", "-C", root, "rev-parse", "--verify", ref],
    { stderr: "pipe" },
  );
  if (result.exitCode !== 0) throw new Error("release tag does not resolve");
  return result.stdout.toString().trim();
}
export function checkSource(
  root: string,
  ref?: string,
  allowUntagged = false,
): string {
  const version = (
    Bun.TOML.parse(readFileSync(resolve(root, "Cargo.toml"), "utf8")) as any
  ).workspace.package.version;
  if (
    typeof version !== "string" ||
    !/^0\.(0|[1-9]\d*)\.(0|[1-9]\d*)$/.test(version)
  )
    throw new Error(
      "Rust release requires an explicit alpha 0.x.y version (ADR 0006)",
    );
  const packages = (
    Bun.TOML.parse(readFileSync(resolve(root, "Cargo.lock"), "utf8")) as any
  ).package;
  for (const name of ["hiero", "hieronymus"]) {
    const versions = packages.filter((item: any) => item.name === name);
    if (versions.length !== 1 || versions[0].version !== version)
      throw new Error(`${name} Cargo.lock version mismatch`);
    const crate = Bun.TOML.parse(
      readFileSync(resolve(root, "crates", name, "Cargo.toml"), "utf8"),
    ) as any;
    if (JSON.stringify(crate.package.version) !== '{"workspace":true}')
      throw new Error(`${name} must inherit the workspace version`);
  }
  if (ref === undefined && allowUntagged) return version;
  if (ref !== `refs/tags/v${version}`)
    throw new Error("release ref must be the exact workspace-version tag");
  if (git(root, "HEAD") !== git(root, `${ref}^{commit}`))
    throw new Error("release tag does not point to checked-out HEAD");
  return version;
}
export async function checkMetadata(
  directory: string,
  version: string,
  channel: string,
): Promise<string> {
  if (!["stable", "dev"].includes(channel))
    throw new Error("unsupported release channel");
  const metadata = JSON.parse(
    readFileSync(resolve(directory, "release.json"), "utf8"),
  );
  const archive = `hieronymus-${version}-${TARGET}.tar.gz`;
  const expected = {
    version,
    target: TARGET,
    channel,
    archive,
    signature: null,
  };
  if (Object.entries(expected).some(([key, value]) => metadata[key] !== value))
    throw new Error(
      "release metadata does not match source/target/channel/waiver",
    );
  if (metadata.sha256 !== (await digest(resolve(directory, archive))))
    throw new Error("release archive checksum mismatch");
  return archive;
}
if (import.meta.main) {
  try {
    const { values } = parseArgs({
      args: Bun.argv.slice(2),
      options: {
        root: { type: "string", default: resolve(import.meta.dir, "..") },
        ref: { type: "string" },
        "allow-untagged": { type: "boolean" },
        "release-dir": { type: "string" },
        channel: { type: "string", default: "stable" },
      },
    });
    const version = checkSource(
      values.root!,
      values.ref,
      values["allow-untagged"],
    );
    const archive = values["release-dir"]
      ? await checkMetadata(values["release-dir"], version, values.channel!)
      : null;
    console.log(
      JSON.stringify({
        version,
        archive,
        tag_verified: values.ref !== undefined,
      }),
    );
  } catch (error) {
    console.error(
      `Rust release refused: ${error instanceof Error ? error.message : "invalid input"}`,
    );
    process.exitCode = 1;
  }
}

File: scripts/check-rust-release.py
#!/usr/bin/env python3
"""Bind Rust release metadata to source, an exact tag, and actual archive bytes."""

from __future__ import annotations

import argparse
import hashlib
import json
import re
import subprocess
import tomllib
from pathlib import Path

TARGET = "x86_64-unknown-linux-gnu"


def digest(path: Path) -> str:
    with path.open("rb") as stream:
        return hashlib.file_digest(stream, "sha256").hexdigest()


def check_source(root: Path, ref: str | None, *, allow_untagged: bool = False) -> str:
    version = tomllib.loads((root / "Cargo.toml").read_text())["workspace"]["package"]["version"]
    if not isinstance(version, str) or not re.fullmatch(r"0\.(0|[1-9]\d*)\.(0|[1-9]\d*)", version):
        raise ValueError("Rust release requires an explicit alpha 0.x.y version (ADR 0006)")
    packages = tomllib.loads((root / "Cargo.lock").read_text())["package"]
    for name in ("hiero", "hieronymus"):
        versions = [item["version"] for item in packages if item["name"] == name]
        if versions != [version]:
            raise ValueError(f"{name} Cargo.lock version mismatch")
        crate = tomllib.loads((root / "crates" / name / "Cargo.toml").read_text())
        if crate["package"]["version"] != {"workspace": True}:
            raise ValueError(f"{name} must inherit the workspace version")
    if ref is None and allow_untagged:
        return version
    if ref != f"refs/tags/v{version}":
        raise ValueError("release ref must be the exact workspace-version tag")
    try:
        head = subprocess.check_output(
            ["git", "-C", str(root), "rev-parse", "HEAD"], text=True
        ).strip()
        tagged = subprocess.check_output(
            ["git", "-C", str(root), "rev-parse", f"{ref}^{{commit}}"], text=True
        ).strip()
    except subprocess.CalledProcessError as error:
        raise ValueError("release tag does not resolve") from error
    if head != tagged:
        raise ValueError("release tag does not point to checked-out HEAD")
    return version


def check_metadata(directory: Path, version: str, channel: str) -> str:
    if channel not in ("stable", "dev"):
        raise ValueError("unsupported release channel")
    metadata = json.loads((directory / "release.json").read_text())
    archive = f"hieronymus-{version}-{TARGET}.tar.gz"
    if (
        any(
            metadata.get(key) != value
            for key, value in {
                "version": version,
                "target": TARGET,
                "channel": channel,
                "archive": archive,
            }.items()
        )
        or "signature" not in metadata
        or metadata["signature"] is not None
    ):
        raise ValueError("release metadata does not match source/target/channel/waiver")
    if metadata.get("sha256") != digest(directory / archive):
        raise ValueError("release archive checksum mismatch")
    return archive


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--root", type=Path, default=Path(__file__).resolve().parents[1])
    parser.add_argument("--ref")
    parser.add_argument(
        "--allow-untagged", action="store_true", help="local candidate only, never publishing"
    )
    parser.add_argument("--release-dir", type=Path)
    parser.add_argument("--channel", default="stable")
    args = parser.parse_args()
    try:
        version = check_source(args.root, args.ref, allow_untagged=args.allow_untagged)
        archive = (
            check_metadata(args.release_dir, version, args.channel) if args.release_dir else None
        )
    except (ValueError, KeyError, OSError) as error:
        parser.exit(1, f"Rust release refused: {error}\n")
    print(
        json.dumps({"version": version, "archive": archive, "tag_verified": args.ref is not None})
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
